### PR TITLE
BHoM_Engine: RunExtensionMethodAsync and TryRunExtensionMethodAsync added

### LINKED ISSUE #3140: BHoM_Engine: add async version of RunExtensionMethod
#### Description:

As in the title, async versions of `RunExtensionMethod` and `TryRunExtensionMethod` methods would be useful.

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #3140

### Test files

Write a simple `public static async` Engine method taking `BHoMObject` and try running it via script uploaded [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FBHoM%5FEngine%2F%233191%2DRunExtensionMethodAsync&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).

### Additional comments

The methods added are to large extent a rewrite of `TryRunExtensionMethod`. However, `async` methods are not allowed to have `out` parameters, therefore I changed the method signature compared to synchronous flavour - happy to discuss if anyone has different/better ideas :+1:

### CODE REVIEW

**Review by alelom — COMMENTED** [review]
This looks good to me. Because this is a "backend" base method, could you please add an Unit test in the BHoM_Engine_Tests solution too? 

It can be a simple test that runs any existing extension method available in the base BHoM_Engine asynchronously, like `Geometry()` for example.

This is where the test project is: https://github.com/BHoM/BHoM_Engine/tree/develop/.ci/unit-tests/Base_Engine_Tests
and this new test should sit in a new `Compute` folder, in a file called like the method you're testing. Just follow the example of the [IsNumeric unit test](https://github.com/BHoM/BHoM_Engine/blob/develop/.ci/unit-tests/Base_Engine_Tests/Query/IsNumeric.cs), for instance.

**Comment by pawelbaran** [comment]
> This looks good to me. Because this is a "backend" base method, could you please add an Unit test in the BHoM_Engine_Tests solution too?
> 
> It can be a simple test that runs any existing extension method available in the base BHoM_Engine asynchronously, like `Geometry()` for example.
> 
> This is where the test project is: https://github.com/BHoM/BHoM_Engine/tree/develop/.ci/unit-tests/Base_Engine_Tests and this new test should sit in a new `Compute` folder, in a file called like the method you're testing. Just follow the example of the [IsNumeric unit test](https://github.com/BHoM/BHoM_Engine/blob/develop/.ci/unit-tests/Base_Engine_Tests/Query/IsNumeric.cs), for instance.

Thanks @alelom for having a look. I was thinking about adding a unit test before I made one up in GH, but the problem is that all Engine methods are synchronous, and it is not possible to call one via `RunExtensionMethodAsync` - you naturally get `MethodName does not contain a definition for 'GetAwaiter'` exception. So I would need to add a dummy async extension method to the engine or call `RunExtensionMethodAsync` inside `RunExtensionMethodAsync`, none of which I was not a big fan of. What do you think?

**Comment by FraserGreenroyd** [comment]
> So I would need to add a dummy async extension method to the engine or call `RunExtensionMethodAsync` inside `RunExtensionMethodAsync`, none of which I was not a big fan of. What do you think?

Haven't read everything through yet, but would agree being against adding a dummy method to the engine for the sake of testing 👍

**Comment by alelom** [comment]
> > This looks good to me. Because this is a "backend" base method, could you please add an Unit test in the BHoM_Engine_Tests solution too?
> > It can be a simple test that runs any existing extension method available in the base BHoM_Engine asynchronously, like `Geometry()` for example.
> > This is where the test project is: https://github.com/BHoM/BHoM_Engine/tree/develop/.ci/unit-tests/Base_Engine_Tests and this new test should sit in a new `Compute` folder, in a file called like the method you're testing. Just follow the example of the [IsNumeric unit test](https://github.com/BHoM/BHoM_Engine/blob/develop/.ci/unit-tests/Base_Engine_Tests/Query/IsNumeric.cs), for instance.
> 
> Thanks @alelom for having a look. I was thinking about adding a unit test before I made one up in GH, but the problem is that all Engine methods are synchronous, and it is not possible to call one via `RunExtensionMethodAsync` - you naturally get `MethodName does not contain a definition for 'GetAwaiter'` exception. So I would need to add a dummy async extension method to the engine or call `RunExtensionMethodAsync` inside `RunExtensionMethodAsync`, none of which I was not a big fan of. What do you think?

It's actually well possible in DotNet -- just wrap the synchronous method in a `Task.Run()` invocation. E.g.
```cs
var result = await Task.Run(() => MySyncMethod());
```

**Comment by alelom** [comment]
> > It's actually well possible in DotNet -- just wrap the synchronous method in a `Task.Run()` invocation. E.g.
> > ```cs
> > var result = await Task.Run(() => MySyncMethod());
> > ```
> 
> Of course, if you have a chance to call `Task.Run` - here we find the method by name and then await the method call:
> 
> https://github.com/BHoM/BHoM_Engine/blob/181d3c2c3a53afb629473a9c71b14e2213d4bf61/BHoM_Engine/Compute/TryRunExtensionMethod.cs#L141

Yeah sure I meant maybe you could go ahead with a dummy method like:

```cs

public async Task<IGeometry> GeometryAsync(IObject obj)
{
    return await Task.Run(() => obj.Geometry());
}

```

**Comment by pawelbaran** [comment]
> Yeah sure I meant maybe you could go ahead with a dummy method like:
> 
> ```cs
> public static async Task<IGeometry> GeometryAsync(this IObject obj)
> {
>     return await Task.Run(() => obj.Geometry());
> }
> ```
> 
> and use that for the invocation in the test. Or something similar

Yeah, but then it needs to be added to a legit engine project to be reflected in the pool parsed by `RunExtenstionMethod`. So we end up with the situation commented by @FraserGreenroyd [here](https://github.com/BHoM/BHoM_Engine/pull/3191#issuecomment-1752660415), or I would need to create a dedicated engine for this one method...

**Comment by alelom** [comment]
> > Yeah sure I meant maybe you could go ahead with a dummy method like:
> > ```cs
> > public static async Task<IGeometry> GeometryAsync(this IObject obj)
> > {
> >     return await Task.Run(() => obj.Geometry());
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and use that for the invocation in the test. Or something similar
> 
> Yeah, but then it needs to be added to a legit engine project to be reflected in the pool parsed by `RunExtenstionMethod`. So we end up with the situation commented by @FraserGreenroyd [here](https://github.com/BHoM/BHoM_Engine/pull/3191#issuecomment-1752660415), or I would need to create a dedicated engine for this one method...

Apologies, it all came back to my mind – this is what I was thinking of having you do. However, I just remembered that I had prepared a "Test Engine" exactly for this purpose – mainly, for testing backend methods like the RunExtensionMethod. I had even added some unit tests, and they were all running fine locally. However, we had immense trouble making them run automatically via CI/CD when BHoMBot attempted to run them via NUnit. I ended up removing them after weeks of trying. 

tl;dr: let's not add this unit test method. Sorry for the long back and forth, I think it was useful to record this exchange anyways.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: unit-tests

**Review by alelom — APPROVED** [review]
Approving based on stated procedure and our conversation above.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge
